### PR TITLE
Monitor supplicant state and inform applications

### PR DESCRIPTION
**Pull request by rado17**
Add a control connection handler for monitoring the supplicant state at real time and notify interested applications.

### CODE REVIEW

File: wpa_supplicant/ctrl_iface_zephyr.c [hunk 1]
```diff
@@ -8,6 +8,123 @@
...
+
+	if (ifname)
+		os_snprintf(levelstr, sizeof(levelstr), "IFNAME=%s <%d>",
+			    ifname, level);
+	else
+		os_snprintf(levelstr, sizeof(levelstr), "<%d>", level);
+
+	sbuf = os_malloc(len);
+	if (sbuf == NULL) {
+		return;
```
Comment (krish2718) [inline wpa_supplicant/ctrl_iface_zephyr.c:73]: log an error

File: wpa_supplicant/ctrl_iface_zephyr.h [hunk 2]
```diff
@@ -16,15 +16,36 @@
 #include "ctrl_iface.h"
 #include "common/wpa_ctrl.h"
 
+/**
+ * struct wpa_ctrl_dst - Data structure of control interface monitors
+ *
+ * This structure is used to store information about registered control
+ * interface monitors into struct wpa_supplicant.
+ */
+struct wpa_ctrl_dst {
```
Comment (krish2718) [inline wpa_supplicant/ctrl_iface_zephyr.h:25]: `dst` might be misleading, we can use other term like `wpa_ctrl_mon` (IIRC, this is what WPAS uses as well)

File: wpa_supplicant/ctrl_iface_zephyr.c [hunk 3]
```diff
@@ -8,6 +8,123 @@
 
 #include "ctrl_iface_zephyr.h"
 
+
+static int wpa_supplicant_ctrl_iface_attach(struct wpa_ctrl_dst **head, int sock)
```
Comment (krish2718) [inline wpa_supplicant/ctrl_iface_zephyr.c:12]: Can we please add `mon` prefix to clearly indicate that these are to monitor events? `wpa_supplicant_ctrl_iface_mon_attach`

File: zephyr/src/supp_events.c [hunk 4]
```diff
@@ -3,10 +3,114 @@
...
+		sscanf(status+strlen("CTRL-EVENT-CONNECTED - Connection to"),
+			MAC_STR_FORMAT, &mac[0], &mac[1], &mac[2], &mac[3], &mac[4], &mac[5]);
+		break;
+
+	case DISCONNECTED:
+		mac = data->disconnected.bssid;
+		sscanf(status+strlen("CTRL-EVENT-DISCONNECTED - bssid="),
+			MAC_STR_FORMAT, &mac[0], &mac[1], &mac[2], &mac[3], &mac[4], &mac[5]);
+		break;
+
```
Comment (krish2718) [inline zephyr/src/supp_events.c:51]: do we really needed these empty lines?

File: zephyr/src/supp_events.h [hunk 5]
```diff
@@ -35,7 +39,85 @@ enum net_event_wpa_supp_cmd {
...
 
-#endif /* __SUPP_EVENTS_H__ */
\ No newline at end of file
+union supp_event_data {
+	struct supp_event_auth_reject {
+		int auth_type;
+		int auth_transaction;
+		int status_code;
+		uint8_t bssid[ETH_ALEN];
+	}auth_reject;
```
Comment (krish2718) [inline zephyr/src/supp_events.h:51]: space after `}`

File: zephyr/src/supp_events.h [hunk 6]
```diff
@@ -35,7 +39,85 @@ enum net_event_wpa_supp_cmd {
 #define NET_EVENT_WPA_SUPP_IFACE_REMOVING					\
 	(_NET_MGMT_WPA_SUPP_EVENT | NET_EVENT_WPA_SUPP_CMD_IFACE_REMOVING)
 
-int send_wifi_mgmt_event(const char *ifname, enum net_event_wifi_cmd event, int status);
+int send_wifi_mgmt_event(const char *ifname, enum net_event_wifi_cmd event, void *status, size_t len);
 int generate_supp_state_event(const char *ifname, enum net_event_wpa_supp_cmd event, int status);
 
-#endif /* __SUPP_EVENTS_H__ */
\ No newline at end of file
+union supp_event_data {
```
Comment (krish2718) [inline zephyr/src/supp_events.h:45]: prefix `mon`

File: zephyr/src/supp_events.c [hunk 7]
```diff
@@ -3,10 +3,114 @@
...
+			&data->bss_removed.id,
+			&mac[0], &mac[1], &mac[2], &mac[3], &mac[4], &mac[5]);
+		break;
+
+	case TERMINATING:
+	case SCAN_STARTED:
+	case SCAN_FAILED:
+	case NETWORK_NOT_FOUND:
+	case NETWORK_ADDED:
+	case NETWORK_REMOVED:
```
Comment (krish2718) [inline zephyr/src/supp_events.c:100]: Should we add a TODO or some comment for these?

File: zephyr/src/supp_events.c [hunk 8]
```diff
@@ -3,10 +3,114 @@
...
+       "CTRL-EVENT-NETWORK-ADDED",
+       "CTRL-EVENT-NETWORK-REMOVED",
+};
+
+static void process_status(struct status_data *event_data, char *status)
+{
+	int event = -1;
+	unsigned char *mac;
+
+	union supp_event_data *data = (union supp_event_data *)event_data->data;
```
Comment (krish2718) [inline zephyr/src/supp_events.c:29]: newline after declaration (can club with above)

**Review by krish2718 — CHANGES_REQUESTED** [review]
I have a PR https://github.com/nrfconnect/sdk-nrf/pull/11899 that rejigs directory and repo structures, so, better to wait to avoid rebasing, we can have this PR in sdk-nrf once my PR is merged.

**Review by sachinthegreen — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]

File: wpa_supplicant/ctrl_iface_zephyr.c [hunk 9]
```diff
@@ -8,6 +8,124 @@
...
+	if (sock < 0 || dst == NULL)
+		return;
+
+	if (ifname)
+		os_snprintf(levelstr, sizeof(levelstr), "IFNAME=%s <%d>",
+			    ifname, level);
+	else
+		os_snprintf(levelstr, sizeof(levelstr), "<%d>", level);
+
+	sbuf = os_malloc(len);
```
Comment (krish2718) [inline wpa_supplicant/ctrl_iface_zephyr.c:71]: Why is this intermediate buffer needed? Can't we just use `conn_msg.msg` itself?

File: wpa_supplicant/ctrl_iface_zephyr.h [hunk 10]
```diff
@@ -16,15 +16,36 @@
...
 struct ctrl_iface_global_priv {
 	struct wpa_global *global;
 	/* 0 - wpa_cli, 1 - ctrl_iface */
 	int sock_pair[2];
+	int mon_sock_pair[2];
+	struct wpa_ctrl_mon *ctrl_dst;
 };
 
+struct conn_msg {
+	char msg[256];
```
Comment (krish2718) [inline wpa_supplicant/ctrl_iface_zephyr.h:49]: better to have a macro for this and also any reason to choose only 256?

Comment (rado17) [reply wpa_supplicant/ctrl_iface_zephyr.h:49]: Length was set to 256 based on a couple of logs whose length was close to 150 bytes.

File: wpa_supplicant/ctrl_iface_zephyr.c (hunk 9 above)
Comment (rado17) [reply wpa_supplicant/ctrl_iface_zephyr.c]: Done

File: wpa_supplicant/ctrl_iface_zephyr.h (hunk 10 above)
Comment (rado17) [reply wpa_supplicant/ctrl_iface_zephyr.h]: Done

**Pull request merged**